cpp/unreal_plugins/UrdfBot/Source/UrdfParser.h

```diff
@@ -0,0 +1,170 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#pragma once
+
+#include <map>
+#include <vector>
+
+#include <Eigen/Dense>
+
+#include <CoreMinimal.h>
+
+class FXmlNode;
+
+struct UrdfJointDesc;
+struct UrdfLinkDesc;
+struct UrdfMaterialDesc;
+
+enum class UrdfGeometryType
+{
+    Invalid = 0,
+    Box = 1,
+    Sphere = 2,
+    Cylinder = 3,
+    Mesh = 4,
+};
+
+enum class UrdfJointType
+{
+    Invalid = 0,
+    Revolute = 1,
+    Continuous = 2,
+    Prismatic = 3,
+    Fixed = 4,
+    Floating = 5,
+    Planar = 6
+};
+
+enum class CalibrationType
+{
+    Invalid = 0,
+    Rising = 1,
+    Falling = 2
+};
+
+struct UrdfGeometryDesc
+{
+    UrdfGeometryType type_;
+    FVector size_ = FVector(0.07); // required by Box, in [m]
+    float radius_ = 0.0352; // required by Sphere, Cylinder, in [m]
+    float length_ = 0.0206; // required by Cylinder, in [m]
+    std::string filename_; // required by Mesh
+    float scale_; // optional for Mesh
+};
+
+struct UrdfMaterialDesc
+{
+    std::string name_;
+
+    FVector4 color_;
+    std::string texture_;
+};
+
+struct UrdfInertialDesc
+{
+    FTransform origin_ = FTransform::Identity;
+    float mass_ = 0.0f;
+    Eigen::Matrix3f inertia_;
+};
+
+struct UrdfVisualDesc
+{
+    std::string name_;
+
+    FTransform origin_ = FTransform::Identity;
+    UrdfGeometryDesc geometry_;
+    UrdfMaterialDesc material_desc_;
+};
+
+struct UrdfCollisionDesc
+{
+    std::string name_;
+
+    FTransform origin_ = FTransform::Identity;
+    UrdfGeometryDesc geometry_;
+};
+
+struct UrdfLinkDesc
+{
+    std::string name_;
+
+    UrdfInertialDesc inertial_desc_;
+    std::vector<UrdfVisualDesc> visual_descs_;
+    std::vector<UrdfCollisionDesc> collision_descs_;
+
+    bool is_root_ = true;
+    std::vector<UrdfLinkDesc*> child_link_descs_ = {};
+    std::vector<UrdfJointDesc*> child_joint_descs_ = {};
+};
+
+struct UrdfJointDesc
```

Comment:
```
struct UrdfJointDesc
{
    std::string name_;

    UrdfJointType type_;
    FTransform origin_ = FTransform::Identity;
    std::string parent_;
    std::string child_;
    FVector axis_ = FVector(1, 0, 0);

    // dynamics
    float damping_ = 0.0f;
    float friction_ = 0.0f;

    // calibration
    CalibrationType calibration_type_ = CalibrationType::Invalid;
    float rising_ = 0.0f;
    float falling_ = 0.0f;

    // limits
    float lower_ = 0.0f;
    float upper_ = 0.0f;
    float effort_ = 0.0f;
    float velocity_ = 0.0f;

    // mimic
    std::string joint_;
    float multiplier_ = 1.0f;
    float offset_ = 0.0f;

    // safety_controller
    float soft_lower_limit_ = 0.0f;
    float soft_upper_limit_ = 0.0f;
    float k_position_ = 0.0f;
    float k_velocity_ = 0.0f;
};

Reply:
I think name_  and type_ should be grouped togethor because both are from attribute while others are from child node. Others are fixed.

Reply:
Fair point. Let's group `name_` and `type_` together.

Reply:
See above.

Reply:
Still need to align equals signs here.

Reply:
Still need to align equals signs here.

Reply:
My bad. Just see the comment and fix them